include/openmp/ThreadState.h

```diff
@@ -0,0 +1,18 @@
+#ifndef FAASM_THREADSTATE_H
+#define FAASM_THREADSTATE_H
+
+#include <openmp/Level.h>
+
+namespace wasm {
+    namespace openmp {
+        // TLS for each OMP thread
+        extern thread_local int thisThreadNumber;
+        extern thread_local OMPLevel *thisLevel;
+
+        void setThreadNumber(int);
+
+        void setThreadLevel(OMPLevel *);
```

Comment:
The lifetime of it is well-defined so ARC is really overkill. I spent a lot of time trying to handle this neatly, as there are issues with initialisation, TLS, etc. I agree though that leaving a pointer to a shared object is never a good idea in the long run, but in the grand scheme of things it is premature to fix this now, as the handling of the OMPLevel is likely to change a bit more in the future

Reply:
What I feared would happen actually happened and when you refactored with the prepare OMP function you leaked and lead to tons of funny things.

I changed to `shared_ptr` because there is no way by manual memory management to get this right without using the stack and the const instantiation with the ternary operator hack that I did before but it pains me because we're paying that price simply because of an initial instantiation issue. I'll get back to this definitely, it'll be a lot cleaner after we do the few refactors I talked about below